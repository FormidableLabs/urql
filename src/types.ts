import { Observer, Source, Subject } from 'wonka';
import { ClientState } from './components';
import { Client } from './lib/client';
import { CombinedError } from './lib/error';

export type RequestOperationType =
  | 'subscription'
  | 'query'
  | 'mutation';

export type SignalOperationType =
  | 'teardown';

/** The type of GraphQL operation being executed. */
export type OperationType = RequestOperationType | SignalOperationType;

/** A Graphql query, mutation, or subscription. */
export interface GraphQLRequest {
  query?: string;
  variables?: object;
}

export type Query = GraphQLRequest;
export type Mutation = GraphQLRequest;
export type Subscription = GraphQLRequest;

/** Additional metadata passed to [exchange]{@link Exchange} functions. */
export interface OperationContext {
  [key: string]: any;
  fetchOptions?: RequestInit;
  url: string;
}

/** A [query]{@link Query} or [mutation]{@link Mutation} with additional metadata for use during transmission. */
export interface RequestOperation extends GraphQLRequest {
  /** Unique identifier of the operation. */
  key: string;
  operationName: RequestOperationType;
  context: OperationContext;
}

/** A signal with additional metadata that informs exchanges about specific events during transmission. */
export interface SignalOperation {
  operationName: SignalOperationType;
  context: OperationContext;
}

export type Operation = RequestOperation | SignalOperation;

// Adapted from: https://github.com/graphql/graphql-js/blob/ae5b163/src/execution/execute.js#L105-L114
export interface ExecutionResult {
  errors?: Error[];
  data?: any;
}

/** Resulting data from an [operation]{@link Operation}. */
export interface ExchangeResult {
  /** The [operation]{@link Operation} which has been executed. */
  operation: RequestOperation;
  /** The data returned from the Graphql server. */
  data?: any;
  /** Any errors resulting from the operation. */
  error?: CombinedError;
}

/** Input parameters for to an Exchange factory function. */
export interface ExchangeInput {
  forward: ExchangeIO;
  client: Client;
}

/** Function responsible for listening for streamed [operations]{@link Operation}. */
export type Exchange = (input: ExchangeInput) => ExchangeIO;

/** Function responsible for receiving an observable [operation]{@link Operation} and returning a [result]{@link ExchangeResult}. */
export type ExchangeIO = (ops$: Source<Operation>) => Source<ExchangeResult>;

/** The arguments for the child function of a connector. */
export interface ChildArgs<MutationDeclarations> {
  /** Whether a dependent GraphQL request is currently being fetched. */
  fetching: ClientState<MutationDeclarations>['fetching'];
  /** Any network or GraphQL errors. */
  error: ClientState<MutationDeclarations>['error'];
  /** The data returned from a GraphQL [query]{@link Query} to the server. */
  data: ClientState<MutationDeclarations>['data'];
  /** A collection of functions for executing pre-specified [mutations]{@link Mutation} */
  mutations: ClientState<MutationDeclarations>['mutations'];
  /** Trigger a fetch of the pre-specified [query]{@link Query}. */
  refetch: (noCache?: boolean) => void;
}
