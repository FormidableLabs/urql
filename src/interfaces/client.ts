import Observable from 'zen-observable-ts';

import { IClientEvent } from '../modules/events';
import { ICache } from './cache';
import { IExchangeResult, IExecutionData } from './exchange';
import { IQuery } from './query';

export interface IClient {
  cache: ICache;
  executeSubscription$(subscriptionObject: IQuery): Observable<IExchangeResult>;
  executeQuery$(
    queryObject: IQuery,
    skipCache: boolean
  ): Observable<IExchangeResult>;
  executeQuery(
    queryObject: IQuery,
    skipCache: boolean
  ): Promise<IExchangeResult>;
  executeMutation$(mutationObject: IQuery): Observable<IExecutionData>;
  executeMutation(mutationObject: IQuery): Promise<IExecutionData>;
  invalidateQuery(queryObject: IQuery): Promise<void>;
  refreshAllFromCache(): void;
  subscribe(callback: (event: IClientEvent) => void): () => void;
}
