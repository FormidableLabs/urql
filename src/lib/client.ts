import {
  filter,
  makeSubject,
  pipe,
  publish,
  share,
  Source,
  Subject,
  subscribe,
  take,
} from 'wonka';

import { defaultExchanges } from '../exchanges';
import { CombinedError } from './error';
import { hashString } from './hash';

import {
  Exchange,
  ExchangeResult,
  GraphQLRequest,
  Mutation,
  Operation,
  OperationContext,
  OperationType,
  Query,
  Subscription,
} from '../types';

/** Options for configuring the URQL [client]{@link Client}. */
export interface ClientOptions {
  /** Target endpoint URL such as `https://my-target:8080/graphql`. */
  url: string;
  /** Any additional options to pass to fetch. */
  fetchOptions?: RequestInit | (() => RequestInit);
  /** An ordered array of Exchanges. */
  exchanges?: Exchange[];
}

/** The URQL application-wide client library. Each execute method starts a GraphQL request and
 returns a stream of results. */
export class Client implements ClientOptions {
  // These are variables derived from ClientOptions
  url: string;
  fetchOptions: RequestInit;
  exchanges: Exchange[];

  // These are internals to be used to keep track of operations
  dispatchOperation: (operation: Operation) => void;
  operations$: Source<Operation>;
  results$: Source<ExchangeResult>;

  constructor(opts: ClientOptions) {
    this.url = opts.url;

    this.fetchOptions =
      typeof opts.fetchOptions === 'function'
        ? opts.fetchOptions()
        : (opts.fetchOptions || {});

    this.exchanges =
      opts.exchanges !== undefined
        ? opts.exchanges
        : defaultExchanges;

    // This subject forms the input of operations; executeOperation may be
    // called to dispatch a new operation on the subject
    const [operations$, nextOperation] = makeSubject<Operation>();
    this.operations$ = operations$;
    this.dispatchOperation = nextOperation;

    // All operations run through the exchanges in a pipeline-like fashion
    // and this observable then combines all their results
    this.results$ = pipeExchanges(this, this.exchanges, this.operations$);
  }

  createOperation = (
    type: OperationType,
    query: GraphQLRequest,
    context: Partial<OperationContext>
  ): Operation => ({
    ...query,
    key: hashString(JSON.stringify(query)),
    operationName: type,
    context: {
      url: this.url,
      fetchOptions: this.fetchOptions,
      ...context
    }
  });

  executeOperation(operation: Operation): Source<ExchangeResult> {
    this.dispatchOperation(operation);
    const isSpecificExchangeResult = filter((res: ExchangeResult) => res.operation.key === operation.key);

    if (operation.operationName === 'mutation') {
      return pipe(this.results$, isSpecificExchangeResult, take(1));
    } else {
      return pipe(this.results$, isSpecificExchangeResult);
    }
  }

  executeQuery = (query: Query, opts: Partial<OperationContext>): Source<ExchangeResult> =>
    this.executeOperation(this.createOperation('query', query, opts));

  executeSubscription = (query: Subscription, opts: Partial<OperationContext>): Source<ExchangeResult> =>
    this.executeOperation(this.createOperation('subscription', query, opts));

  executeMutation = (query: Mutation, opts: Partial<OperationContext>): Source<ExchangeResult> =>
    this.executeOperation(this.createOperation('mutation', query, opts));
}

/** Create pipe of exchanges */
const pipeExchanges = (
  client: Client,
  exchanges: Exchange[],
  operation$: Source<Operation>
): Source<ExchangeResult> => {
  /** Recursively pipe to each exchange */
  const callExchanges = (value: Source<Operation>, index: number = 0) => {
    if (index >= exchanges.length) {
      return value;
    }

    const currentExchange = exchanges[index];
    return currentExchange({
      forward: val => callExchanges(val, index + 1),
      client,
    })(value);
  };

  return share(callExchanges(operation$, 0));
};
