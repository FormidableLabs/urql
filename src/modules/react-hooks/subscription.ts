// @ts-ignore
import { useCallback, useMemo, useState, useEffect, useContext } from 'react';
import { context } from '../../components/context';
import { ISubscriptionHook } from '../../interfaces/index';
import Client from '../client';

export function useSubscription<Data>(
  query: string,
  variables?: Object
): ISubscriptionHook<Data> {
  const COMPARISON = [query, variables];

  const client: Client = useContext(context);
  const [fetching, setFetching] = useState(true);
  const [error, setError] = useState(null);
  const [data, setData] = useState(null);
  const [loaded, setLoaded] = useState(false);

  let connection;

  // Fetch the query
  useEffect(() => {
    if (connection) {
      connection.unsubscribe();
      connection = null;
    }

    connection = client.executeSubscription$({ query, variables }).subscribe({
      complete: () => {
        this.subscriptionSub = null;
      },
      error: e => {
        this.subscriptionSub = null;
        setError(e);
        setFetching(false);
      },
      next: result => {
        setData(result.data || null);
        setError(result.error || null);
        setFetching(false);
        setLoaded(true);
      },
    });
  }, COMPARISON);

  return { data, loaded, error, fetching };
}
