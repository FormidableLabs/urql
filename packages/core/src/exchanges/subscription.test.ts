import { print } from 'graphql';
import {
  empty,
  publish,
  fromValue,
  pipe,
  Source,
  take,
  toPromise,
} from 'wonka';
import { Client } from '../client';
import {
  subscriptionOperation,
  subscriptionResult,
  queryOperation,
} from '../test-utils';
import { OperationResult } from '../types';
import { subscriptionExchange, SubscriptionForwarder } from './subscription';

it('should return response data from forwardSubscription observable', async () => {
  const exchangeArgs = {
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = operation => {
    expect(operation.query).toBe(print(subscriptionOperation.query));
    expect(operation.variables).toBe(subscriptionOperation.variables);
    expect(operation.context).toEqual(subscriptionOperation.context);

    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next(subscriptionResult);
        });

        return { unsubscribe };
      },
    };
  };

  const data = await pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription })(exchangeArgs),
    take(1),
    toPromise
  );

  expect(data).toMatchSnapshot();
  expect(unsubscribe).toHaveBeenCalled();
});

it('should tear down the operation if the source subscription ends', async () => {
  const reexecuteOperation = jest.fn();
  const unsubscribe = jest.fn();

  const exchangeArgs = {
    forward: () => empty as Source<OperationResult>,
    client: { reexecuteOperation: reexecuteOperation as any } as Client,
  };

  const forwardSubscription: SubscriptionForwarder = () => ({
    subscribe(observer) {
      observer.complete();
      return { unsubscribe };
    },
  });

  pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription })(exchangeArgs),
    publish
  );

  await Promise.resolve();

  expect(unsubscribe).not.toHaveBeenCalled();
  expect(reexecuteOperation).toHaveBeenCalled();
});

it('should allow for query operations with all operations enabled', async () => {
  const exchangeArgs = {
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = operation => {
    expect(operation.query).toBe(print(queryOperation.query));
    expect(operation.variables).toBe(queryOperation.variables);
    expect(operation.context).toEqual(queryOperation.context);

    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next(subscriptionResult);
        });

        return { unsubscribe };
      },
    };
  };

  const data = await pipe(
    fromValue(queryOperation),
    subscriptionExchange({
      enableAllOperations: true,
      forwardSubscription,
    })(exchangeArgs),
    take(1),
    toPromise
  );

  expect(data).toMatchSnapshot();
  expect(unsubscribe).toHaveBeenCalled();
});

it('should allow for selective operation skipping with all operations enabled', async () => {
  const exchangeArgs = {
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = operation => {
    expect(operation.query).toBe(print(queryOperation.query));
    expect(operation.variables).toBe(queryOperation.variables);
    expect(operation.context).toEqual(queryOperation.context);

    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next(subscriptionResult);
        });

        return { unsubscribe };
      },
    };
  };

  const data = await pipe(
    fromValue(queryOperation),
    subscriptionExchange({
      enableAllOperations: true,
      forwardSubscription,
    })(exchangeArgs),
    take(1),
    toPromise
  );

  expect(data).toMatchSnapshot();
  expect(unsubscribe).toHaveBeenCalled();
});

it('should allow for selective operation skipping with all operations enabled', async () => {
  const exchangeArgs = {
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = () => {
    expect(true).toBeFalsy();

    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next(subscriptionResult);
        });

        return { unsubscribe };
      },
    };
  };

  const data = pipe(
    fromValue(queryOperation),
    subscriptionExchange({
      enableAllOperations: true,
      forwardSubscription,
      skipOperation: op => op.key === queryOperation.key,
    })(exchangeArgs)
  );

  expect(data).toMatchSnapshot();
  expect(unsubscribe).toHaveBeenCalledTimes(0);
});
