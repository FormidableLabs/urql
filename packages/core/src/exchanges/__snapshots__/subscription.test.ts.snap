// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow for query operations with all operations enabled 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "operationName": "query",
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 124,
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`should allow for selective operation skipping with all operations enabled 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "operationName": "query",
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 124,
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`should allow for selective operation skipping with all operations enabled 2`] = `[Function]`;

exports[`should return response data from forwardSubscription observable 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 4,
    "operationName": "subscription",
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "subscribeToUser",
          },
          "operation": "subscription",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "user",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "user",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 106,
        "start": 0,
      },
    },
    "variables": Object {
      "user": "colin",
    },
  },
}
`;
